upstream: Idle time out for upstream connections

ramaraochavali opened the pull request:
*Description*:  Adds Idle connection timeout support for upstream connections
*Risk Level*: Low 
*Testing*: Automated tests
*Docs Changes*: https://github.com/envoyproxy/data-plane-api/pull/513
*Release Notes*:
>If this change is user impacting you **must** add a release note to
[RAW_RELEASE_NOTES.md](RAW_RELEASE_NOTES.md). Thank you! Please write in N/A if
there are no release notes.

ramaraochavali commented:
@mattklein123 @htuch initial implementation. Can you please take a look?
Still
- Need to add tests
- Need to merge data-plane-api and use it's value here.

htuch commented:
@ggreenway can you take first pass on this? Thanks.

ggreenway commented on source/common/http/http1/conn_pool.cc:
```diff
@@ -81,6 +81,8 @@ ConnectionPool::Cancellable* ConnPoolImpl::newStream(StreamDecoder& response_dec
     ready_clients_.front()->moveBetweenLists(ready_clients_, busy_clients_);
     ENVOY_CONN_LOG(debug, "using existing connection", *busy_clients_.front()->codec_client_);
     attachRequestToClient(*busy_clients_.front(), response_decoder, callbacks);
+    //reset idle timer
```
delete this comment; the code is self-explanatory here

ggreenway commented on source/common/http/http1/conn_pool.cc:
```diff
@@ -94,6 +96,8 @@ ConnectionPool::Cancellable* ConnPoolImpl::newStream(StreamDecoder& response_dec
     // If we have no connections at all, make one no matter what so we don't starve.
     if ((ready_clients_.size() == 0 && busy_clients_.size() == 0) || can_create_connection) {
       createNewConnection();
+      //reset idle timer
```
delete comment

ggreenway commented on source/common/http/http1/conn_pool.cc:
```diff
@@ -281,8 +288,8 @@ ConnPoolImpl::PendingRequest::~PendingRequest() {
 ConnPoolImpl::ActiveClient::ActiveClient(ConnPoolImpl& parent)
     : parent_(parent),
       connect_timer_(parent_.dispatcher_.createTimer([this]() -> void { onConnectTimeout(); })),
+      idle_timer_(parent_.dispatcher_.createTimer([this]() -> void { onIdleTimeout(); })),
```
You should only create the timer if a timeout is configured

ggreenway commented on source/common/http/http1/conn_pool.h:
```diff
@@ -80,6 +80,21 @@ class ConnPoolImpl : Logger::Loggable<Logger::Id::pool>, public ConnectionPool::
+    void resetIdleTimer() {
+      if (idle_timer_) {
+        idle_timer_->disableTimer();
+        idle_timer_.reset();
```
I think you probably shouldn't delete the timer here; just disable it, so that it can be re-used on the next idle period. And rename the function to disableIdleTimer.

If you delete it here, that means you only get to check for idleness after the first request, but not after subsequent requests (because you never re-create it)

ggreenway commented on source/common/http/http2/conn_pool.h:
```diff
@@ -45,6 +45,21 @@ class ConnPoolImpl : Logger::Loggable<Logger::Id::pool>, public ConnectionPool::
+    void resetIdleTimer() {
+      if (idle_timer_) {
+        idle_timer_->disableTimer();
+        idle_timer_.reset();
```
Same as in http1 re: reset()

ggreenway commented on source/common/http/http2/conn_pool.cc:
```diff
@@ -213,11 +224,15 @@ void ConnPoolImpl::onStreamReset(ActiveClient& client, Http::StreamResetReason r
   }
+
+  //Start the idle timer
+  client.startIdleTimer();
```
Is the entire connection idle at this point?  Can there be other active streams?

ggreenway commented:
I think a good next step on this PR is to write some integration tests (for both http 1.1 and 2), and get those passing, in addition to getting the existing tests to pass.

mattklein123 commented:
> At least for http2, it looks like this isn't correctly handling multiple streams on the same connection. It may be more appropriate to put this on the codec_impl instead of on the conn_pool.

@ggreenway yeah I think either would work. Potentially doing this on the codec client would allow it to be easily used for both HTTP/1.1 and HTTP/2. We still definitely need to plumb through the right stats though.

JonathanO commented on source/common/http/http1/conn_pool.cc:
```diff
@@ -81,6 +81,7 @@ ConnectionPool::Cancellable* ConnPoolImpl::newStream(StreamDecoder& response_dec
     ready_clients_.front()->moveBetweenLists(ready_clients_, busy_clients_);
     ENVOY_CONN_LOG(debug, "using existing connection", *busy_clients_.front()->codec_client_);
     attachRequestToClient(*busy_clients_.front(), response_decoder, callbacks);
+    busy_clients_.front()->disableIdleTimer();
```
I'm not familiar enough with this code to be sure, but is there a potential race here where the idle timer can expire after the request is attached, but before the timer is disabled? What will happen in that case?

ramaraochavali replied:
That is a good point. Should I move the disableTimer before attachRequestToClient? Would it help?

ramaraochavali commented:
@ggreenway @htuch I probably need some guidance on simulating the idletimeout behavior in http2 integration tests. Can you point me to relevant calls in the test framework that can simulate this behavior?

alyssawilk commented:
@ramaraochavali  I'd suggest checking an existing timeout integration test (Ratelimit has one)

Good principles are configuring a timeout short enough it won't cause the tests to be slow (1s or less) and having a stat you can check to make sure the right thing happened (you didn't run afoul of some other timeout) and so Envoy users with configured timeouts know how often timeouts occur.  For now you can just hard-code timeout on and once your  api change is approved you can add a config modifier to set a low value   But having even a test with a handful of TODOs in http_integration.cc and run in the HTTP and H2 integration tests will be a great start towards making sure the feature works as expected :-)

mattklein123 commented:
@ramaraochavali per @ggreenway should we move this timeout functionality into codec_client? It seems potentially simpler to do it in one place that will work for either h1 or h2?

ggreenway commented:
@ramaraochavali Ping me when you have the tests added and passing; I'm going to hold off on further review until then. I think the tests will make it much clearer if the direction of implementation is working out.

ramaraochavali commented:
@alyssawilk thanks for your suggestion on integration test. I tried to write today on the similar lines but I got stuck because my changes are in `CodecClient` while integration tests use a separate `CodecClient`- so the changes are not getting invoked and hence not able to see the metrics.
I have tested this manually and it works, after timeout the metric gets incremented and client is closed. But I am not sure now how to automate this.
@htuch @mattklein123 any suggestions on automating this integration test?

ramaraochavali commented:
@htuch @mattklein123 an update, just to prove that it works, i have added `CodecClientCallback` and added the logic to kick the idleTimeout there. Added a basic test to see if that works. If this is ok, i will add further tests. Otherwise let me know if there is any better way to write integration test for  this.

htuch commented on source/common/http/codec_client.cc:
```diff
@@ -37,13 +41,17 @@ void CodecClient::deleteRequest(ActiveRequest& request) {
     codec_client_callbacks_->onStreamDestroy();
   }
+  if (numActiveRequests() == 0) {
+    enableIdleTimer();
```
Should this also be enabled initially? Or do we assume we will always have an inflight request at codec creation?

ramaraochavali replied:
I went with that assumption and in tests that I did it is true also. I added that condition to ensure that it is enabled only when there are no active streams.

htuch commented on test/integration/http_integration.cc:
```diff
@@ -839,6 +842,44 @@ void HttpIntegrationTest::testEnvoyProxying100Continue(bool continue_before_upst
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_200", 1);
+
+  // do not send any requests and validate if idle time out kicks in.
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_idle_timeout", 1);
```
This style of testing seems reasonable to me.

ramaraochavali commented:
@htuch That was one of my questions to @mattklein123 in the beginning fo this PR. How should long running streams like our xDS work with this timeout? are you suggesting that if there is a stream opened but no activity - we should just kill it after some x hours?

ramaraochavali commented:
@ggreenway added tests..can you PTAL?

mattklein123 commented:
@ramaraochavali @htuch for long running stream timeouts please see: https://github.com/envoyproxy/envoy/issues/1778. We will implement data frame timeouts for both rx and tx.

IMO this feature is really mostly about HTTP/1.1 and racing with upstream idle timeout so I think what we are doing here is appropriate.

ggreenway commented on RAW_RELEASE_NOTES.md:
```diff
@@ -8,6 +8,7 @@ will make it substantially easier for the releaser to "linkify" all of the relea
 final version.
 
 ## 1.6.0
+* Added support for specifying idle timeout for upstream clusters.
```
Added support for specifying idle timeout for upstream HTTP connections.

ggreenway commented on source/common/http/codec_client.cc:
```diff
@@ -22,7 +23,10 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
+  if (host != nullptr && host_->cluster().idleTimeout().valid()) {
+    idle_timer_ = dispatcher_.createTimer([this]() -> void { onIdleTimeout(); });
+    idle_timeout_ = host_->cluster().idleTimeout().value();
+  }
```
Call enableIdleTimer() here? Or is the CodecClient always created with an active stream?

ramaraochavali replied:
yes. the CodecClient always created with an active stream

ggreenway commented on source/common/http/codec_client.h:
```diff
@@ -121,10 +122,30 @@ class CodecClient : Logger::Loggable<Logger::Id::client>,
+  }
+
+  void disableIdleTimer() {
+    if (idle_timer_) {
```
if (idle_timer_ != nullptr)

ggreenway commented on source/common/http/codec_client.h:
```diff
@@ -121,10 +122,30 @@ class CodecClient : Logger::Loggable<Logger::Id::client>,
+  }
+
+  void enableIdleTimer() {
+    if (idle_timer_) {
```
if (idle_timer_ != nullptr)

ggreenway commented on source/common/http/codec_client.h:
```diff
@@ -121,10 +122,30 @@ class CodecClient : Logger::Loggable<Logger::Id::client>,
   ClientConnectionPtr codec_;
   Network::ClientConnectionPtr connection_;
   Upstream::HostDescriptionConstSharedPtr host_;
+  Event::Dispatcher& dispatcher_;
```
You don't need this reference. This is only used in the constructor to create the timer.

ggreenway commented on test/integration/http2_integration_test.cc:
```diff
@@ -195,6 +195,110 @@ TEST_P(Http2IntegrationTest, Trailers) { testTrailers(1024, 2048); }
+}
+
+// Interleave two requests and responses and make sure that idle timeout is handled correctly.
+void Http2IntegrationTest::simultaneousRequestsWithIdleTimeout() {
```
No need for separate function that is the only thing called by a test. Just inline this entire thing into the single test that is calling it.

The reason some of the other tests above call testTrailers() is because they're calling the same function with different parameters in each case.

ggreenway commented on test/integration/http2_integration_test.cc:
```diff
@@ -195,6 +195,110 @@ TEST_P(Http2IntegrationTest, Trailers) { testTrailers(1024, 2048); }
+  // Validate that timeout is set and is as per expectation.
+  bool has_idle_timeout =
+      config_helper_.bootstrap().static_resources().clusters(0).has_common_http_protocol_options();
+  EXPECT_TRUE(has_idle_timeout);
```
No need to validate this; this is just validating that previous test code in this function worked.

ggreenway commented on test/integration/http2_integration_test.cc:
```diff
@@ -195,6 +195,110 @@ TEST_P(Http2IntegrationTest, Trailers) { testTrailers(1024, 2048); }
 
 TEST_P(Http2IntegrationTest, TrailersGiantBody) { testTrailers(1024 * 1024, 1024 * 1024); }
 
+TEST_P(Http2IntegrationTest, IdleTimoutBasic) { testIdleTimeoutBasic(); }
```
Please add a 1-2 sentence comment above each test saying what it tests. We're trying to do this for all new tests (and adding them to old tests whenever they are touched).

ggreenway commented on test/integration/http2_integration_test.cc:
```diff
@@ -195,6 +195,110 @@ TEST_P(Http2IntegrationTest, Trailers) { testTrailers(1024, 2048); }
+
+  // Do not send any requests and validate idle timeout kicks in after both the requests are done.
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_idle_timeout", 1);
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_total", 0);
```
This condition is always true: the counters are unsigned, so it is always >= 0. Did you mean to check for equality to 0?

ramaraochavali replied:
Good point. I can not check equality to 0 because, cx_total would be incremented any way. So there is no value of this check and will remove.

ramaraochavali commented:
@ggreenway addressed all the comments. PTAL.

mattklein123 commented on test/integration/http2_integration_test.cc:
```diff
@@ -195,6 +195,104 @@ TEST_P(Http2IntegrationTest, Trailers) { testTrailers(1024, 2048); }
+
+  // Cleanup both downstream and upstream
+  codec_client_->close();
+  fake_upstream_connection1->close();
```
Are the close calls needed here? Can we just move this up above line 286 and wait for disconnect due to idle timeout?

mattklein123 replied:
This one still needs action?

mattklein123 commented on test/integration/http_integration.h:
```diff
@@ -30,9 +30,22 @@ class IntegrationCodecClient : public Http::CodecClientProd {
+  void onStreamDestroy() override {
+    num_active_requests_--;
+    if (num_active_requests_ == 0) {
+      enableIdleTimer();
```
Do we want idle timeout behavior in the integration codec client? I think probably not?

ramaraochavali replied:
Initially it was required. Now with the change i am making to enable it in the constructor itself it may not be required. will test and remove it.

mattklein123 commented on test/integration/http_integration.cc:
```diff
@@ -839,6 +843,92 @@ void HttpIntegrationTest::testEnvoyProxying100Continue(bool continue_before_upst
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_200", 1);
+
+  // Do not send any requests and validate if idle time out kicks in.
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_idle_timeout", 1);
```
Can we verify that fake upstream connections are closed by waiting for closure? Same elsewhere.

ramaraochavali commented on source/common/http/codec_client.cc:
```diff
@@ -22,7 +23,10 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
   ENVOY_CONN_LOG(debug, "connecting", *connection_);
   connection_->connect();
-
+  if (host != nullptr && host_->cluster().idleTimeout()) {
```
It happens in unit tests. Looks like the mocked tests are not passing/constructing the host. That is why I have added this.

ramaraochavali replied:
ok.i got where nullptr is being passed in tests. will correct it

ramaraochavali commented on source/common/http/codec_client.h:
```diff
@@ -121,10 +122,29 @@ class CodecClient : Logger::Loggable<Logger::Id::client>,
   Network::ClientConnectionPtr connection_;
   Upstream::HostDescriptionConstSharedPtr host_;
+  Event::TimerPtr idle_timer_;
+  std::chrono::milliseconds idle_timeout_;
```
i made it optional but could not make it const because the host sometimes in null in mock tests.

ramaraochavali commented:
@mattklein123 addressed the review comments. PTAL.

mattklein123 commented on source/common/http/codec_client.cc:
```diff
@@ -23,6 +25,10 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
   ENVOY_CONN_LOG(debug, "connecting", *connection_);
   connection_->connect();
 
+  if (host_->cluster().idleTimeout()) {
```
nit: You can just look at `idle_timeout_` here.

mattklein123 commented on source/common/http/codec_client.cc:
```diff
@@ -12,8 +12,10 @@ namespace Envoy {
 namespace Http {
 
 CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
```
I think there may be places in which codec clients get destroyed as part of deferred deletion. Can you make sure the idle timer is disabled/destroyed when the codec client is closed by any mechanism? Please add a test that covers this also (close client when timer is active and no active requests).

mattklein123 commented:
Per new release note policy please add a release note with any applicable feature docs to https://github.com/envoyproxy/data-plane-api/blob/master/docs/root/intro/version_history.rst. Thank you!

ramaraochavali commented:
@mattklein123  added the logic to `disableIdleTimer` in `close` as well. Also added the test to verify this behaviour. Regarding the release notes, I will follow-up a PR in data-plane-api to remove the "unhide" for the new proto messages and adding it to release notes once this is merged.

mattklein123 commented on source/common/http/codec_client.cc:
```diff
@@ -23,27 +25,39 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
 CodecClient::~CodecClient() {}
 
-void CodecClient::close() { connection_->close(Network::ConnectionCloseType::NoFlush); }
+void CodecClient::close() {
```
Instead of disabling here, you should just reset() the timer when you receive either a local/remote close here: https://github.com/envoyproxy/envoy/blob/master/source/common/http/codec_client.cc#L50. This will ensure the timer never rearms after any type of close.

ramaraochavali commented:
@mattklein123 moved the `disableTimer` to `connectionEvent` callback with `reset` as the timer would no longer be needed. OSX test rate limiting integration test failed not related to this change, and now passed.  I think this should be good to go. PTAL.

mattklein123 commented on source/common/http/codec_client.cc:
```diff
@@ -74,6 +84,7 @@ void CodecClient::onEvent(Network::ConnectionEvent event) {
           connected_ ? StreamResetReason::ConnectionTermination
                      : StreamResetReason::ConnectionFailure);
     }
+    disableIdleTimer(true);
```
You need to move this before you start closing active requests, around line 80. Also, just do `idle_timer_.reset()`. You don't need the extra parameter to `disableIdleTimer`. Please also add a test for the case where the client is closed with both active requests and active idle timer to https://github.com/envoyproxy/envoy/blob/master/test/common/http/codec_client_test.cc.

mattklein123 commented on test/mocks/upstream/cluster_info.cc:
```diff
@@ -29,6 +29,7 @@ MockClusterInfo::MockClusterInfo()
       resource_manager_(new Upstream::ResourceManagerImpl(runtime_, "fake_key", 1, 1024, 1024, 1)) {
 
   ON_CALL(*this, connectTimeout()).WillByDefault(Return(std::chrono::milliseconds(1)));
+  ON_CALL(*this, idleTimeout()).WillByDefault(Return(idle_timeout_));
```
Having the member variable `idle_timeout_` in the mock doesn't do anything since `Return` captures by value. Either use `ReturnPointee` or just `Return` a temporary here.

ramaraochavali commented:
@mattklein123  added tests in codec_client_test to validate those scenarios. PTAL.

mattklein123 commented:
@ramaraochavali please check compile error. I will review rest of change in a bit. Thanks!

mattklein123 commented on include/envoy/upstream/upstream.h:
```diff
@@ -283,6 +283,7 @@ class PrioritySet {
   COUNTER  (upstream_cx_http2_total)                                                               \
   COUNTER  (upstream_cx_connect_fail)                                                              \
   COUNTER  (upstream_cx_connect_timeout)                                                           \
+  COUNTER  (upstream_cx_idle_timeout)                                                              \
```
Sorry forgot this before. Can you do a follow up doc PR to add this stat to here: https://www.envoyproxy.io/docs/envoy/latest/configuration/cluster_manager/cluster_stats ?

ramaraochavali replied:
will add

mattklein123 commented on test/common/http/common.h:
```diff
@@ -31,7 +32,7 @@ class CodecClientForTest : public Http::CodecClient {
 
   void raiseGoAway() { onGoAway(); }
-
+  Event::TimerPtr idleTimer() { return std::move(idle_timer_); }
```
`Event::Timer* idleTimer() { return idle_timer_.get(); }`

ramaraochavali replied:
if I change this, it never returns nullptr even after the reset call and the following check in codec_client_test fails.
  EXPECT_EQ(client_->idleTimer(), nullptr);
So reverted it like this. Let me know if there is any other better way for this check.

mattklein123 replied:
Something doesn't make sense. The move call to return the local variable is not correct. If the above is returning nullptr I think it's likely doing so by accident somehow. If you are trying to look at the member variable state, you should do it by pointer, not move.

Another way to handle is to just remove this function entirely, and expect `disableTimer()`. This is generally what is done elsewhere. I think that would be cleaner anyway.

ramaraochavali replied:
ok. you are correct. I found the problem. The close method on CodecClient is not actually invoking onEvent and hence idletimer is not disabled in that case. It is a great catch. I have added that code in close method as well and now it works.

mattklein123 commented on test/integration/http_integration.cc:
```diff
@@ -422,7 +422,6 @@ void HttpIntegrationTest::testDrainClose() {
                                                                {":authority", "host"}},
                                        *response_);
   response_->waitForEndStream();
-  codec_client_->waitForDisconnect();
```
? Why this deletion?

ramaraochavali replied:
The actual disconnection is happening. but waitForDisconnect is returning false.. probably because of other dispatcher events...that is why removed it.
   [2018-03-20 00:30:19.289][3380685][debug][connection] source/common/network/connection_impl.cc:134] [C85] closing socket: 1
[2018-03-20 00:30:19.289][3380685][debug][client] source/common/http/codec_client.cc:81] [C85] disconnect. resetting 0 pending requests
[2018-03-20 00:30:19.289][3381124][trace][connection] source/common/network/connection_impl.cc:386] [C86] socket event: 1
[2018-03-20 00:30:19.289][3381124][trace][connection] source/common/network/connection_impl.cc:424] [C86] read ready
[2018-03-20 00:30:19.289][3381124][trace][connection] source/common/network/raw_buffer_socket.cc:25] [C86] read returns: 0
test/integration/http_integration.cc:142: Failure
Value of: disconnected_
  Actual: false
Expected: true

mattklein123 replied:
I don't understand how your change could have effected this test? There is no idle timeout here? I'm guessing a previous test (one you added) is not cleaning up properly?

ramaraochavali replied:
@mattklein123 found the problem. Some of the integration tests are using MockClusterInfo for creating http connection. Since now the MockClusterInfo returns the valid idle timeout value, idle timeout is happening and client is closing. The LocalClose event is generated which is not handled by the IntegrationCodecClient. I have increased the idle timeout value and reverted the waitForDisconnect. Now all tests are passing.

mattklein123 replied:
I think a better/safer fix here would be to mack the mock return the default optional value, so that there is no timeout by default. Do you mind making that change?

ramaraochavali replied:
i will make that change. But the codec_client_test needs idletimeout set..will have to figure out how to set the value there

ramaraochavali replied:
ok. added MockIdleTimeEnabledClusterInfo which overrides that method and returns the value. This can be used in future for other tests if needed.

ramaraochavali replied:
i have reverted this. but added below. I can move it to the correct positions.

mattklein123 commented on test/integration/http_integration.cc:
```diff
@@ -848,6 +847,128 @@ void HttpIntegrationTest::testEnvoyProxying100Continue(bool continue_before_upst
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_total", 1);
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_200", 1);
+
+  // Close the Codec Client and validate Idle timer is also disabled.
```
What is this test testing exactly? It's a little unclear?

ramaraochavali replied:
agreed..this test is not required..will remove it

mattklein123 commented:
Option 1. Override mock using ON_CALL in your test. Option 2. Bring back
member, use ReturnPointee, and change member.

On Mar 19, 2018 10:33 PM, "Rama Chavali" <notifications@github.com> wrote:

> *@ramaraochavali* commented on this pull request.
> ------------------------------
>
> In test/integration/http_integration.cc
> <https://github.com/envoyproxy/envoy/pull/2714#discussion_r175659165>:
>
> > @@ -422,7 +422,6 @@ void HttpIntegrationTest::testDrainClose() {
>                                                                 {":authority", "host"}},
>                                         *response_);
>    response_->waitForEndStream();
> -  codec_client_->waitForDisconnect();
>
> i will make that change. But the codec_client_test needs idletimeout
> set..will have to figure out how to set the value there
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/2714#discussion_r175659165>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGA17GAaW2UNTADyndezCbwnrzfJo0Qeks5tgJS4gaJpZM4SbzeE>
> .
>

ramaraochavali commented:
@mattklein123 yes, that is what i am doing. adding a `MockIdleTimeEnabledClusterInfo` class and using that here. so that it can be used else where in future

mattklein123 commented on source/common/http/codec_client.cc:
```diff
@@ -23,27 +25,40 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
-void CodecClient::close() { connection_->close(Network::ConnectionCloseType::NoFlush); }
+void CodecClient::close() {
+  connection_->close(Network::ConnectionCloseType::NoFlush);
+  disableIdleTimer();
```
I don't understand why this was added back here. What you had before was correct. Close should absolutely raise events, unless the connection was already closed somehow. There is likely some bug in the test code if only the mock test was failing. Can you please investigate?

ramaraochavali replied:
Ok. The reason the test code using a MockConnection which is not raising the callbacks. The connnection callbacks need to be raised specifically and rest of the tests do the same. Raised them and it works. PTAL.

mattklein123 commented on test/common/http/codec_client_test.cc:
```diff
@@ -128,6 +131,82 @@ TEST_F(CodecClientTest, DisconnectBeforeHeaders) {
+  // Close the client and validate idleTimer is reset
+  EXPECT_EQ(client_->numActiveRequests(), 0);
+  client_->close();
+  connection_cb_->onEvent(Network::ConnectionEvent::LocalClose);
```
OK I see the issue, it looks like the CodecClient tests don't use the default connection mock handlers for raising events which is kind of a shame. We should add a TODO here to fix.

mattklein123 commented on test/common/http/codec_client_test.cc:
```diff
@@ -128,6 +131,82 @@ TEST_F(CodecClientTest, DisconnectBeforeHeaders) {
+  // When we get a local close with an active request validate idleTimer is reset after client close
+  EXPECT_CALL(callbacks, onResetStream(StreamResetReason::ConnectionTermination));
+  connection_cb_->onEvent(Network::ConnectionEvent::Connected);
+  connection_cb_->onEvent(Network::ConnectionEvent::LocalClose);
```
Same here with TODO. Optimally, we should also call close() on the client to make the test more realistic.

mattklein123 commented on test/common/http/common.h:
```diff
@@ -31,7 +32,7 @@ class CodecClientForTest : public Http::CodecClient {
   }
 
   void raiseGoAway() { onGoAway(); }
-
```
nit: del newline, add newline between functions and `destroy_cb_`

mattklein123 commented on test/integration/http2_integration_test.cc:
```diff
@@ -195,6 +195,99 @@ TEST_P(Http2IntegrationTest, Trailers) { testTrailers(1024, 2048); }
+  // Do not send any requests and validate idle timeout kicks in after both the requests are done.
+  fake_upstream_connection1->waitForDisconnect();
+  fake_upstream_connection2->waitForDisconnect();
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_idle_timeout", 1);
```
Shouldn't this be 2?

ramaraochavali replied:
yes..good catch..will change
